clusterloader2/docs/experiments.md

```diff
@@ -0,0 +1,44 @@
+# clusterloader2 - experiment rollout
+
+In this doc any change to the behavior of clusterloader2 is referred as an "experiment".
+
+## Motivation
+
+clusterloader2 is a tool used a by all scalability (performance) tests. Tests compile clusterloader2 at [HEAD](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-scalability/sig-scalability-release-blocking-jobs.yaml#L121). Introducing breaking changes to clusterloader2 will stop tests from passing at all. While for smaller and faster tests, this is not a huge costs (unless this happens on the weekend, see https://github.com/kubernetes/perf-tests/pull/586), this can be costly in rarely run, expensive tests (e.g. [ci-kubernetes-e2e-gce-scale-performance](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-scalability/sig-scalability-release-blocking-jobs.yaml#L44)).
+
+For this reason, all new changes features added to clusterloader2, shall be gated and rolled out gradually. This way we can minimize blast radius of breaking changes. We can even remove some kinds of them to happen at all, as they should be caught at presubmit time and not allowed to merge at all.
+
+## Experiment rollout
+
+### General principles
+
+1. If for some reason it doesn't make sense to enable an experiment in particular test job indicate the reason clearly in the PR (e.g. measuring kube-proxy metrics in kubemark can be skipped as kube-proxy is not running there).
+
+1. Allow at least 24h between previous step to ensure the experiment is stable. We don't want to block PR to k/k because of a flaky experiment in clusterloader2.
+
+1. Check with sig-scalability whether there is a regression in the test you want to enable experiment for. We don't want new features in clusterloader2 to interfere with regression' debugging.
+
+1. For tests not listed below it's fine to enable experiments
+
+### Step-by-step process
+
+_Each step should be a separate PR_
+
+1. Enable experiment in [perf-tests presubmits](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-scalability/sig-scalability-presubmit-jobs.yaml#L267) and [ci-kubernetes-e2e-gci-gce-scalability](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-scalability/sig-scalability-release-blocking-jobs.yaml#L98)
```

Comment:
So overall the steps are

1. perf-tests presubmits and gce-100 CI
2. k/k presubmits
3. kubemark CI
4. gce-scale CI

Questions:
1. Can we split the first step and have a gce-100 CI as a separate step?
2. What about GKE tests using CL2?

Reply:
ad. 1) Yes, we can. However enabling an experiment only for perf-test presubmit doesn't add much value IMO. It's only executed when PR for this repo is opened, which is not-frequent. gce-100 CI should give us enough data points and on the other hands, if things go south we can revert an experiment quickly. 

ad. 2) GKE tests are not release-blocking and used by PRs. IMO you can enable experiment at any point you like, assuming you will take care if you break them. Do we want more formalized approach?

Reply:
> ad. 1) Yes, we can. However enabling an experiment only for perf-test presubmit doesn't add much value IMO.

I disagree - it is critical. It would actually prevent the issue that happened last week.
The flow in my opinion should be:
- change test-infra config for perf-tests presubmit
- introduce the env-var in perf-tests - that would automatically test whether the config is correct
- enable it everywhere else

Regarding gradual rollout - I'm not sure how much it buys us. We generally should be easily able to assess the performance impact of the chances in clusterloader from the code. So as long as we enable first the experiments in "non-blocking" mode, I don't think it actually provides much value.

Reply:
Wojtek, I think the import part of that sentence is the word **only**. By no means I wanted to say, that updating perf-tests presubmit jobs is not important. It is! But on the other hand it gives us little signal, as those jobs are not run continuously. 

As mentioned in another comment - we already rollout changes gradually. This doc simply tries to codify this practice in an opinionated way. One reason we don't want to enable experiment everywhere at once is regressions. It happen much more often that we have problems with gce5000, while gce100 is doing just fine. To not interfere with debugging, we wait with enabling in gce5000, while the new code is doing just fine in gce100. While waiting for resolving the regression, we shouldn't allow new code to k/k that would break with the experiment. I'd rather have this reflected in the process, than making exceptions to it few times a quarter. 

Also, I disagree, that's easy to access the performance impact. It seems that our tests are prone to even few more API calls (like the current situation with extra lists from kubelet).

Reply:
>  It is! But on the other hand it gives us little signal, as those jobs are not run continuously.

My main motivation now is to avoid breaking the test because of bad push. And that solves this problem.

> It happen much more often that we have problems with gce5000, while gce100 is doing just fine. To not interfere with debugging, we wait with enabling in gce5000, while the new code is doing just fine in gce100.

If it's doing just fine in 100-node, what's the point with waiting for it?
I actually claim that we will know how it behaves on 100-node cluster just from presubmit.

> Also, I disagree, that's easy to access the performance impact. It seems that our tests are prone to even few more API calls (like the current situation with extra lists from kubelet).

Listing all objects of a given type (which was the case from kubelet) is a know thing that will cause issues - it is not surprising at all.
I agree that there may be some subtle things, but I actually claim that we won't catch them by gradual rollout too (they won't be visible in 100-node clusters).

Reply:
1. A good addition to this discussion is the issue from this week, when innocent change by Maciek broke all our tests and kubernetes presubmit. Gradual rollout in this case would reduce the blast radius only to our test only. Personally I think it's worth add a little work on us to prevent breaking k/k presubmit. That's afterall one of my main motivation for this doc - to avoid breakages of k/k presubmit because of bugs on our side. 

2. re: waiting: we sometimes want to wait, because there is an ongoing regression, not because there is something wrong with experiment. This has happened few time last quarter, that we wait with experiments, until regressions are resolved. 

3. I guess we both know that listing object can be source of issues. But it's no obvious for everyone. But anyway, this was just an example. I can imagine other, more subtle problems, which may not be easy to spot.

Reply:
> A good addition to this discussion is the issue from this week, when innocent change by Maciek broke all our tests and kubernetes presubmit. Gradual rollout in this case would reduce the blast radius only to our test only. Personally I think it's worth add a little work on us to prevent breaking k/k presubmit. That's afterall one of my main motivation for this doc - to avoid breakages of k/k presubmit because of bugs on our side.

That my motivation too (and actually that's the only thing that I really would like to achieve).
And to do that you need to:
1. first submit all the code to actually use it (including test-infra)
2. then send out PR ti perf-tests to start using  that env-var/config-file or sth to perf-tests (test-definition) so that the presubmit on perf-tests repo will run against full change
3. once that passes, (assuming our presubmit is identical that the one in k/k repo, which hopefuly is), we know we won't break that and 100-node tests (because presubmit in perf-tests repo passed), and we can enable it everywhere else

My main point here is: it's point 1 and 2 being done in the correct order that will prevent those issues - not the gradual rollout itself.

Reply:
I've changed the content according to our offline discussion. 

I've also remove step with enabling change in Kubemark.